agent/configuration-build-derive/src/derive.rs

```diff
@@ -0,0 +1,433 @@
+        let ty = field.ty.clone();
+        let vis = field.vis.clone();
+        Some(quote! {
+            #(#doc)*
+            #vis #field_ident: model::ConfigValue<#ty>,
+        })
+    });
+    // Create the setters for each field, one for typed values and one for templated strings
+    let setters = data.fields.iter().filter_map(|field| {
+        let doc = field.attrs.iter().filter(|v| {
+            v.parse_meta()
+                .map(|meta| meta.path().is_ident("doc"))
+                .unwrap_or(false)
+        });
+        let field_name = match field.ident.as_ref() {
+            Some(ident) => ident.to_string(),
+            None => return None,
+        };
+        let field_ident = Ident::new(&field_name, Span::call_site());
+        let template_field_name = format!("{}{}", field_name, "_template");
+        let template_ident = Ident::new(&template_field_name, Span::call_site());
+        let ty = field.ty.clone();
+        Some(quote! {
+            #(#doc)*
+            #[inline(always)]
+            pub fn #field_ident<T>(&mut self, #field_ident: T) -> &mut Self
+            where
+            T: Into<#ty>{
+                self.#field_ident = model::ConfigValue::Value(#field_ident.into());
+                self
+            }
+
+            #[inline(always)]
+            pub fn #template_ident<S1,S2>(&mut self, resource: S1, field: S2) -> &mut Self
+            where
+            S1: Into<String>,
+            S2: Into<String>,
+            {
+                self.#field_ident = model::ConfigValue::TemplatedString(format!("${{{}.{}}}", resource.into(), field.into()));
+                self
+            }
+        })
+    });
+    let fns = quote! {
+        #(#setters)*
+    };
+
+    // Add the build function to the builders.
+    let build = match crd_type.as_str() {
+        "Test" => {
+            quote! {
+                #[derive(Debug, Default, serde::Serialize)]
+                pub struct #build_ident{
+                    #(#fields)*
+                    depends_on: Vec<String>,
+                    resources: Vec<String>,
+                    labels: std::collections::BTreeMap<String,String>,
+                    image: Option<String>,
+                    image_pull_secret: Option<String>,
+                    secrets: std::collections::BTreeMap<String, model::SecretName>,
+                    retries: Option<u32>,
+                    keep_running: Option<bool>,
+                    capabilities: Vec<String>,
+                }
+
+                impl #build_ident{
+                    pub fn new() -> #build_ident {
+                        Default::default()
+                    }
+
+                    #fns
+
+                    pub fn depends_on<S1>(&mut self, depends_on: S1) -> &mut Self
+                    where
+                    S1: Into<String> {
+                        self.depends_on.push(depends_on.into());
+                        self
+                    }
```

Comment:
What's the reasoning for having both `fn set_<field>` and `fn <field>`? Can we just stick with one? It seems like `fn <field>` is for adding one more to the collection and `fn set_<field>` is setting the entire collection. Do we need both?

Reply:
I think it's important to have both. `set_` is not as clean looking as the other ones, but if a user is reusing the builder to create several similar test/resources a `set_` function is the only way to clear out the value.